File: src/main/java/it/scoppelletti/spaceship/gradle/android/AppPlugin.java
/*
 * Copyright (C) 2020-2021 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.gradle.android;

import java.util.Objects;
import javax.annotation.Nonnull;
import com.android.build.gradle.AppExtension;
import it.scoppelletti.spaceship.gradle.SpaceshipPlugin;
import it.scoppelletti.spaceship.gradle.android.model.SpaceshipAppExtension;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.logging.Logger;
import org.gradle.api.logging.Logging;

/**
 * Plug-in for Android Apps.
 *
 * @since 1.0.0
 */
@SuppressWarnings("unused")
public class AppPlugin implements Plugin<Project> {
    private static final Logger myLogger = Logging.getLogger(AppPlugin.class);

    @Override
    public void apply(@Nonnull Project project) {
        AppExtension androidExt;

        if (project.getPlugins().hasPlugin(SpaceshipPlugin.class)) {
            myLogger.info("Plugin {} already applied.", SpaceshipPlugin.class);
        } else {
            myLogger.info("Applying plugin {}.", SpaceshipPlugin.class);
            project.getPluginManager().apply(SpaceshipPlugin.class);
        }

        androidExt = Objects.requireNonNull(
                project.getExtensions().findByType(AppExtension.class),
                () -> String.format("Extension %1$s not found.",
                        AppExtension.class));

        project.getExtensions().create(SpaceshipAppExtension.NAME,
                SpaceshipAppExtension.class, project);

        androidExt.getApplicationVariants().all(variant -> {
            AndroidAppTools tools;
            AndroidAppTaskNames taskNames;

            taskNames = new AndroidAppTaskNames(variant);
            tools = new AndroidAppTools(project, variant, taskNames);

            tools.generateCredits();
        });
    }
}

File: src/main/kotlin/it/scoppelletti/spaceship/gradle/android/AppPlugin.kt
/*
 * Copyright (C) 2020-2022 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.gradle.android

import com.android.build.api.artifact.MultipleArtifact
import com.android.build.api.variant.ApplicationAndroidComponentsExtension
import com.android.build.api.variant.ApplicationVariant
import it.scoppelletti.spaceship.gradle.android.tasks.CreditsTask
import org.gradle.api.Plugin
import org.gradle.api.Project

/**
 * Plugin for Android Apps.
 *
 * @since 1.0.0
 */
public abstract class AppPlugin: Plugin<Project> {

    override fun apply(project: Project) {
        project.extensions.findByType(
            ApplicationAndroidComponentsExtension::class.java)?.let { ext ->
            ext.onVariants(ext.selector().all()) { variant ->
                onVariant(project, variant)
            }
        } ?: run {
            project.logger.error("""Extension
                |${ApplicationAndroidComponentsExtension::class.java} not
                |found.""".trimMargin().replace('\n', ' '))
        }
    }

    /**
     * Completes a variant.
     *
     * @param project Project.
     * @param variant Variant.
     */
    @Suppress("UnstableApiUsage")
    private fun onVariant(project: Project, variant: ApplicationVariant) {
        val tools = CreditsTools.create(project)
        val creditsTask = tools.createCreditsTask(variant)
        variant.artifacts.use(creditsTask)
            .wiredWith(CreditsTask::outputDir)
            .toAppendTo(MultipleArtifact.ASSETS)
    }
}